File: backend/src/models/Voucher.model.js
import mongoose from 'mongoose';

const voucherRowSchema = new mongoose.Schema({
    sn: Number,
    dc: { type: String, enum: ['C', 'D'] },
    account: { type: String, required: true },
    accountId: { type: mongoose.Schema.Types.ObjectId, ref: 'Account' },
    balance: { type: Number, default: 0 },
    debit: { type: Number, default: 0 },
    credit: { type: Number, default: 0 },
    modeOfPayment: { type: String, enum: ['Cash', 'Bank', 'Cheque'], default: 'Cash' },
    docType: String,
    chqDocNo: String,
    chqDocDate: Date,
    shortNarration: String,
    remark: String
});

const voucherSchema = new mongoose.Schema({
    recordType: { type: String, enum: ['Payment', 'Receipt', 'Journal', 'Contra', 'Debit', 'Credit'], required: true },
    billSeries: { type: String, required: true },
    billNo: { type: mongoose.Schema.Types.Mixed, required: true }, // can be string or num
    date: { type: Date, required: true },
    gstApplicable: { type: String, default: 'Not Applicable' },
    inputTaxCredit: { type: String, default: 'Not Applicable' },
    rcm: { type: String, default: 'Not Applicable' },
    rows: [voucherRowSchema],
    totalDebit: { type: Number, default: 0 },
    totalCredit: { type: Number, default: 0 },
    remarks: String,
    reffSeries: String,
    reffPurchaseNo: String,
    odrSeries: String,
    vouchNo: String,
    reffPurchaseDate: Date,
    other: String
}, { timestamps: true });

export default mongoose.model('Voucher', voucherSchema);

File: lib/features/vouchers/data/models/voucher_model.dart
// ignore_for_file: invalid_annotation_target
import 'package:freezed_annotation/freezed_annotation.dart';

part 'voucher_model.freezed.dart';
part 'voucher_model.g.dart';

@freezed
class VoucherModel with _$VoucherModel {
  const factory VoucherModel({
    @JsonKey(name: '_id') String? id,
    @Default('Payment') String recordType,
    @Default('') String billSeries,
    @Default('') dynamic billNo, // can be string or num
    String? date,
    @Default('Not Applicable') String gstApplicable,
    @Default('Not Applicable') String inputTaxCredit,
    @Default('Not Applicable') String rcm,
    @Default([]) List<VoucherRowModel> rows,
    @Default(0.0) double totalDebit,
    @Default(0.0) double totalCredit,
    String? remarks,
    String? reffSeries,
    String? reffPurchaseNo,
    String? odrSeries,
    String? vouchNo,
    String? reffPurchaseDate,
    String? other,
    String? createdAt,
    String? updatedAt,
  }) = _VoucherModel;

  factory VoucherModel.fromJson(Map<String, dynamic> json) =>
      _$VoucherModelFromJson(json);
}

@freezed
class VoucherRowModel with _$VoucherRowModel {
  const factory VoucherRowModel({
    @JsonKey(name: '_id') String? id,
    @Default(0) int sn,
    @Default('D') String dc, // 'C' or 'D'
    @Default('') String account,
    @Default('') String accountId,
    @Default(0.0) double balance,
    @Default(0.0) double debit,
    @Default(0.0) double credit,
    @Default('Cash') String modeOfPayment,
    String? docType,
    String? chqDocNo,
    String? chqDocDate,
    String? shortNarration,
    String? remark,
  }) = _VoucherRowModel;

  factory VoucherRowModel.fromJson(Map<String, dynamic> json) =>
      _$VoucherRowModelFromJson(json);
}
